crates/ruff_python_semantic/src/scope.rs

```diff
@@ -189,11 +200,23 @@ impl<'a> Scopes<'a> {
     }
 
     /// Pushes a new scope and returns its unique id
-    pub fn push_scope(&mut self, kind: ScopeKind<'a>) -> ScopeId {
+    pub fn push_scope(&mut self, kind: ScopeKind<'a>, parent: ScopeId) -> ScopeId {
         let next_id = ScopeId::try_from(self.0.len()).unwrap();
-        self.0.push(Scope::local(next_id, kind));
+        self.0.push(Scope::local(next_id, parent, kind));
         next_id
     }
+
+    /// Returns an iterator over all [`ScopeId`] ancestors, starting from the given [`ScopeId`].
+    pub fn ancestors(&self, scope: ScopeId) -> impl Iterator<Item = ScopeId> + '_ {
+        std::iter::successors(Some(scope), |&scope| self[scope].parent)
+    }
+
+    /// Returns an iterator over all [`Scope`] ancestors, starting from the given [`ScopeId`].
+    pub fn ancestor_scopes(&self, scope: ScopeId) -> impl Iterator<Item = &Scope> + '_ {
+        std::iter::successors(Some(&self[scope]), |&scope| {
+            scope.parent.map(|parent| &self[parent])
+        })
```

Comment:
A bit odd to have two methods here, but sort of convenient...

Reply:
Nit: I would flip the two so that `ancestors` returns an iterator of `Scope`s and `ancestor_ids` returns the ids only. 

The `ids` operator may not be required because the `id` can be accessed by using `Scope.id()`.